### ISSUE TITLE
twisted.web.http.Request.addCookie should handle both bytes and str arguments

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8067]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8067|
|Type|defect|
|Created|2015-10-26 04:18:40Z|
|Branch|https://github.com/twisted/twisted/tree/addCookie-str-8067|

A potential thing to be mindful of is http://stackoverflow.com/questions/4400678/http-header-should-use-what-character-encoding -- the current implementation enforces ASCII, but users may wish to put random bytes into the cookie if they so wish, but enforcing ASCII is a good default.

<details><summary>Searchable metadata</summary>

```
trac-id__8067 8067
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__Twisted_16_1 Twisted 16.1
branch__branches_addCookie_str_8067 branches/addCookie-str-8067
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1445833120053716 1445833120053716
changetime__1457687067873439 1457687067873439
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|SamyCookie commented|
|-|-|

I agree with this ticket, 'addCookie' with str parameters provoke [#8077](https://github.com/twisted/twisted/issues/8077) issue in Python 3.

see my proposed solution there: https://github.com/twisted/twisted/compare/trunk...SamyCookie:8067

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

Headers now operates on str and bytes, so I think it's reasonable to allow this to handle both properly.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

(In [46936]) Branching to addCookie-str-8067.

**Comment by twisted-trac** [comment]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|SamyCookie commented|
|-|-|

I have not checked if it is working with str, but I suppose it is.
Some tests are now failing with my changes, I will propose a better patch.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

This should be a duplicate of [#8077](https://github.com/twisted/twisted/issues/8077)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

@SamyCookie I've got a patch in the works, I should have it up for review when I'm back from food :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

There's some implementation details that are similar to http://twistedmatrix.com/trac/ticket/8129 -- so I'm going to do something like that (handling both bytes and unicode). Unicode is fine for 99% of cases, and the bytes fallback allows people to send binary garbage or other encodings fine.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

+1 for more Unicode love :)

**Comment by twisted-trac** [comment]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|SamyCookie commented|
|-|-|

@hawkowl Ok, I'll wait your solution for review. I was trying last hour to propose mine, but Python 2/3 encodings make me crazy.

+1 either for a bytes/str/unicode agnostic solution !

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

Okay, I'm reasonably happy with this. Builders spun, please review.

**Comment by twisted-trac** [comment]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|SamyCookie commented|
|-|-|

It was pretty much that I want to write (but I failed to write an _ensureBytes function correctly :/ ), hence I am totally OK with the logic.

Minor: just saw L1540 and L1595, you forgot to reuse 'expectedCookieValue' variable.

Tests are fine but seems redundant. Is it possible to factorize them ?

**Comment by twisted-trac** [comment]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|SamyCookie commented|
|-|-|

The proposed patch is correctly fixing [#8077](https://github.com/twisted/twisted/issues/8077) wrong behaviour to me.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban commented|
|-|-|

Here is my quick review

I remember that someone else tried to fix this and we hit a problem when the max_age was passed as int.

Major comment:

1 . I think that this branch also needs a removal note as from now on max_age will no longer accept an int... it used to work with an int.

Minor comment:

1. Maybe we can have a helper ``` self.assertCookies([expectedCookieValue], req) ``` which should reduce the duplicate code for req.write('') and then split and then filter the lines.

2. Do we need the special handling of `None` in `_ensureBytes` ? It will fail anyway.

-------

Leaving this for review 

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Please see my the previous comments, mine and from SamyCookie


Since SamyCookie also commented about code duplication in test I would take it as a major issue.

Please address the comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46977]) Merge addCookie-str-8067: Make twisted.web.http.Request.addCookie take bytes and unicode arguments

Author: hawkowl
Reviewers: SamyCookie, adiroiban
Fixes: [#8067](https://github.com/twisted/twisted/issues/8067)

**Comment by twisted-trac** [comment]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|SamyCookie commented|
|-|-|

Thanks for fixing this, cookie sessions should be OK now with Python 3 ([#8077](https://github.com/twisted/twisted/issues/8077)).
